pkg/controller/cni/cilium_reconciler.go

```diff
@@ -0,0 +1,245 @@
+package cni
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	eksacluster "github.com/aws/eks-anywhere/pkg/cluster"
+	"github.com/aws/eks-anywhere/pkg/constants"
+	"github.com/aws/eks-anywhere/pkg/controller"
+	"github.com/aws/eks-anywhere/pkg/controller/serverside"
+	"github.com/aws/eks-anywhere/pkg/executables"
+	"github.com/aws/eks-anywhere/pkg/networking/cilium"
+)
+
+const defaultRequeueTime = time.Second * 10
+
+type CiliumReconciler struct {
+	Client  client.Client
+	Log     logr.Logger
+	tracker *remote.ClusterCacheTracker
+}
+
+func (cr *CiliumReconciler) Reconcile(ctx context.Context, cluster *anywherev1.Cluster, capiCluster *clusterv1.Cluster, specWithBundles *eksacluster.Spec) (controller.Result, error) {
+	cr.Log.Info("Getting remote client", "client for cluster", capiCluster.Name)
+	key := client.ObjectKey{
+		Namespace: capiCluster.Namespace,
+		Name:      capiCluster.Name,
+	}
+	remoteClient, err := cr.tracker.GetClient(ctx, key)
+	if err != nil {
+		return controller.Result{}, err
+	}
+
+	cr.Log.Info("Applying CNI")
+	ciliumDS := &v1.DaemonSet{}
+	ciliumDSName := types.NamespacedName{Namespace: "kube-system", Name: cilium.DaemonSetName}
+	err = remoteClient.Get(ctx, ciliumDSName, ciliumDS)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			cr.Log.Info("Deploying Cilium DS")
+			helm := executables.NewHelm(executables.NewExecutable("helm"), executables.WithInsecure())
+
+			ci := cilium.NewCilium(nil, helm)
```

Comment:
I think we should build this beforehand and inject it in the constructor, probably as an interface

And probably avoid having that nil in the constructor. That makes me think we need a smaller entity than `Cilium`